### ISSUE TITLE
Smaller last batch skews Train/EvalResult

### ISSUE THREAD

**Issue by huyvnphan** [PyTorchLightning/pytorch-lightning#2924]
## 🚀 Feature
Supported weighted average for Result
### Motivation
The last batch of dataloader is usually smaller than a regular batch. Hence simply average the statistic of all batches will skew the epoch statistic.
_For example_    
```
result.log('accuracy', 0.7) # batch size 256
result.log('accuracy', 0.8) # batch size 256
result.log('accuracy', 0.9) # batch size 10 (last batch)
```
Wrong epoch accuracy =` (0.7 + 0.8 + 0.9) / 3`
Correct epoch accuracy = `(256 * 0.7 + 256 * 0.8 + 10 * 0.9) / ( 256 + 256 + 10)`

### Pitch

Support input for weight, or batch size in Result
`result.log('accuracy', 0.7, batch_size) 
`
### Alternatives

I don't see any alternatives other than switch back the the old way of using training/validation_epoch_end and do the weighted average manually there.

**Comment by justusschock** [comment]
@huyvnphan thanks for the issue, you're totally right. I'll try to tackle this issue :)

Unfortunately I cannot give you an ETA so far :)

**Comment by williamFalcon** [comment]
good catch! 

There are a few options for this:

1. you can add a reduce function yourself that can do a weighted average result(key, val, reduce_fx=torch.mean).  torch.mean is the current default but you can implement a weighted average if you want. 

2. we can switch the default to a weighted average. 

3. maybe the better solution is to track the batch sizes internally (which we already do since we have the batch during this step), and do this automatically.
